poseidon2-starky/src/stark.rs

```diff
@@ -1 +1,282 @@
+use starky::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use starky::stark::Stark;
+use starky::vars::{StarkEvaluationTargets, StarkEvaluationVars};
+use std::marker::PhantomData;
+use zkhash::fields::goldilocks::FpGoldiLocks;
 
+const M4: [[usize; 4]; 4] = [
+    [5, 7, 1, 3], //
+    [4, 6, 1, 1], //
+    [1, 3, 5, 7], //
+    [1, 1, 4, 6],
+];
+
+fn scalar_to_fe<F: RichField + Extendable<D>, const D: usize, FE, const D2: usize, PF: PrimeField>(
+    scalar: PF,
+) -> FE
+where
+    FE: FieldExtension<D2, BaseField = F>,
+{
+    FE::from_canonical_u64(
+        F::from_noncanonical_biguint(BigUint::from_bytes_le(&scalar.into_bigint().to_bytes_le()))
+            .to_canonical_u64(),
+    )
+}
+
+// linear layer (degree = 1)
+fn matmul_external8_constraints<
+    F: RichField + Extendable<D>,
+    const D: usize,
+    FE,
+    P,
+    const D2: usize,
+>(
+    state: &[P; 8],
+) -> [P; 8]
+where
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    assert_eq!(STATE_SIZE, 8);
+    let mut out = [P::ZEROS; 8];
+
+    for i in 0..8 {
+        for j in 0..4 {
+            out[i] = if i < 4 {
+                out[i] + state[j] * FE::from_canonical_usize(M4[i][j])
+            } else {
+                out[i] + state[j + 4] * FE::from_canonical_usize(M4[i - 4][j])
+            }
+        }
+    }
+
+    let mut stored = [P::ZEROS; 4];
+    for i in 0..4 {
+        stored[i] = out[i] + out[4 + i];
+    }
+    for i in 0..8 {
+        out[i] = out[i] + stored[i % 4];
+    }
+
+    out
+}
+
+fn matmul_internal8_constraints<
+    F: RichField + Extendable<D>,
+    const D: usize,
+    FE,
+    P,
+    const D2: usize,
+>(
+    state: &[P; 8],
+) -> [P; 8]
+where
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    assert_eq!(STATE_SIZE, 8);
+    let mut out = [P::ZEROS; 8];
+    let mut sum = P::ZEROS;
+
+    for i in 0..STATE_SIZE {
+        sum = sum + state[i];
+    }
+
+    for i in 0..STATE_SIZE {
+        out[i] = state[i] * scalar_to_fe::<F, D, FE, D2, FpGoldiLocks>(MAT_DIAG8_M_1[i]);
+        out[i] = out[i] + sum;
+    }
+
+    out
+}
+
+// degree: 1
+fn add_rc_constraints<F: RichField + Extendable<D>, const D: usize, FE, P, const D2: usize>(
+    state: &[P; 8],
+    r: usize,
+) -> [P; 8]
+where
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    assert_eq!(STATE_SIZE, 8);
+    let mut out = [P::ZEROS; 8];
+
+    for i in 0..8 {
+        out[i] = state[i] + scalar_to_fe::<F, D, FE, D2, FpGoldiLocks>(RC8[r][i]);
+    }
+
+    out
+}
+
+// degree: 7
+fn sbox_p_constraints<F: RichField + Extendable<D>, const D: usize, FE, P, const D2: usize>(
+    state: &P,
+) -> P
+where
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    assert_eq!(STATE_SIZE, 8);
+    let mut out = P::ONES;
+
+    for _ in 0..SBOX_DEGREE {
+        out = out * *state;
```

Comment:
Can you maybe consider using slightly less `asterix rich` syntax ?

Reply:
Done